docs/general.md

```diff
@@ -534,4 +534,82 @@ add 8,r0,r1
+3. **Integration with EraVM**:
+    - Retrieves operational code from the `Decommitter` using a hash from the server.
+    - Uses its heap to dynamically interface with the server and process transaction data systematically.
+4. **Contract Code Management**: Retrieves or defaults contract code based on specific conditions, ensuring correct handling during contract calls.
+
+#### Bootloader Interaction with `decommit_code_hash`
+
+The `decommit_code_hash` function from EraVM manages the contract code hash, which is critical for executing contracts during block construction. The Bootloader interacts with this function to access the contract code hash and execute the contract.
+
+1. **Contract Identification**: It identifies the system's deployer contract address, generating a `storage_key` to access the contract's code in storage:
+    ```rust
+    let deployer_system_contract_address = Address::from_low_u64_be(DEPLOYER_SYSTEM_CONTRACT_ADDRESS_LOW as u64);
+    let storage_key = StorageKey::new(deployer_system_contract_address, address_into_u256(address));
+    ```
+
+2. **Reading and Initializing Contract Code**: It reads the code associated with the contract address. If no code exists, it initializes storage with a default value:
+    ```rust
+    let code_info = match storage.storage_read(storage_key)? {
+        Some(code_info) => code_info,
+        None => {
+            storage.storage_write(storage_key, U256::zero())?;
+            U256::zero()
+        }
+    };
+    ```
+3. **Checking Construction Status**: It verifies if the contract is constructed by checking specific flags within the code:
+    ```rust
+    let is_constructed = match code_info_bytes[1] {
+        IS_CONSTRUCTED_FLAG_ON => true,
+        IS_CONSTRUCTED_FLAG_OFF => false,
+        _ => return Err(EraVmError::IncorrectBytecodeFormat);
+    };
+    ```
+4. **Handling Different Contract Versions**: It processes different contract versions (e.g., EraVM contracts, EVM blobs) based on their state:
+    ```rust
+    code_info_bytes = match code_info_bytes[0] {
+        CONTRACT_VERSION_FLAG => {
+            if is_constructed == is_constructor_call {
+                try_default_aa.ok_or(StorageError::KeyNotPresent)?
+            } else {
+                code_info_bytes
+            }
+        }
+        BLOB_VERSION_FLAG => try_default_aa.ok_or(StorageError::KeyNotPresent)?,
+        _ if code_info == U256::zero() => try_default_aa.ok_or(StorageError::KeyNotPresent)?,
+        _ => return Err(EraVmError::IncorrectBytecodeFormat),
+    };
+    ```
+5. **Returning the Code Hash**: Finally, it prepares the code hash that the Bootloader will use for executing the contract:
+    ```rust
+    code_info_bytes[1] = 0;
+    Ok(U256::from_big_endian(&code_info_bytes))
+    ```
+This hash is critical, as it corresponds to the contract code executed during block construction.
```

Comment:
I would also add a section explaining how the hooks work, as it seems that there’s no documentation available on this topic. There is a brief explanation about the [hook pointer](https://docs.zksync.io/zk-stack/components/zksync-evm/bootloader#vm-hook-pointers) in the bootloader that might be useful.

To briefly explain how it works: basically, the bootloader uses a function call, [`setHook`](https://github.com/matter-labs/era-contracts/blob/8670004d6daa7e8c299087d62f1451a3dec4f899/system-contracts/bootloader/bootloader.yul#L3932), that writes to a specific part of the memory reserved to indicate that a hook is present. The VM executes this store to memory and [checks if it has been written in that specific part](https://github.com/matter-labs/vm2/blob/9a38900d7af9b1d72b47ce3be980e77c1239a61d/src/instruction_handlers/heap_access.rs#L117). If it has, the VM stores the hook type (there are several types) and the instruction to execute after the hook is handled. The VM will indicate that execution has stopped, and this will be [handled on the zksync-era side](https://github.com/matter-labs/zksync-era/blob/10674620d1a04333507ca17b9a34ab3cb58846cf/core/lib/multivm/src/versions/vm_fast/vm.rs#L111). Once the hook is handled, execution resumes, with the VM using the previously saved instruction pointer.

I’m not sure if it’s necessary to detail all the different parts involved in the hook processing, but that’s a general overview.

Reply:
Thank you! I have added the section, very helpful your resources and explanation.